### DISCUSSION TITLE
Создание списка из дат за диапазон с шагом в 7 дней

### DISCUSSION THREAD

**Discussion by an220022** [totumonline/totum-ru-issues-and-discussions#18] [Как это сделать?] [answered]
Как создать список из дат за определенный диапазон с периодичностью в 7 дней?

**Comment by ttm-alexeypolunin** [comment, marked as answer]
**Подход такой:**

1. Делаем список основу при помощи [listNumberRange](https://docs.totum.online/functions#fn-listNumberRange) с тем шагом, который необходим.

2. При помощи [listReplace](https://docs.totum.online/functions#fn-listReplace) меняем цифры на даты через [dateAdd](https://docs.totum.online/functions#fn-dateAdd), те мы добавляем количество дней к стартовой дате.

**Код:**

```totum

= : listReplace(list: $range; action: $dateadd; value: "value")

		range: listNumberRange(min: 0; max: 100; step: 10)
	
	dateadd: dateAdd(date: $datestart; days: $#value; format: "Y-m-d")
		datestart: $#nd

```


**Результат:**

```json
[
  "2021-10-05",
  "2021-10-15",
  "2021-10-25",
  "2021-11-04",
  "2021-11-14",
  "2021-11-24",
  "2021-12-04",
  "2021-12-14",
  "2021-12-24",
  "2022-01-03"
]
```

**Reply by an220022** [reply]
Через переменную и while можно было бы тоже самое сделать?

**Reply by ttm-alexeypolunin** [reply]
`while` для итерационных задач возможен, но не слишком изящен для Totum — тк нам придется использовать переменную для хранения обновляемых на каждой итерации значений:

```totum

= : while(limit: 10; action: $varadd)
	varadd: var(name: "dates_list"; value: $updatedlist)
		updatedlist: listAdd(list: $var; item: $next_date)
				var: var(name: "dates_list"; default: $#lc)
			next_date: dateAdd(date: $date; days: 10; format: "Y-m-d")
				date: if(condition: $var = $#lc; then: $#nd; else: $var[$last])
					last: listCount(list: $var) - 1

``` 

**Что мы здесь делаем:** выполняем 10 итераций на каждой добавляя в переменную `dates_list` новое значение с датой.

Важным является вызов `var` — мы не можем вызвать как `$#var` тк при первом вызове этой переменной не будет и код вернет ошибку. А вызов через функцию [var](https://docs.totum.online/functions#fn-var) для этого случая имеет параметр `default` в который мы запихиваем `$#lc`.

На каждой итерации кроме первой мы для вычисления последующей даты берем последнюю дату из списка — вычисляем мы ее как `listCount() - 1`.

Еще есть возможность выполнить эту же задачу через [рекурсивный вызов](https://docs.totum.online/syntax#sx-recursive-call). Я завтра докину такой пример в эту ветку.
